queries/sun_position.py

```diff
@@ -0,0 +1,623 @@
+"""
+
+    Greynir: Natural language processing for Icelandic
+
+    Example of a grammar query processor module.
+
+    Copyright (C) 2021 Miðeind ehf.
+
+       This program is free software: you can redistribute it and/or modify
+       it under the terms of the GNU General Public License as published by
+       the Free Software Foundation, either version 3 of the License, or
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module handles queries regarding time of sunrise/sunset.
+
+"""
+from typing import Dict, List, Tuple, Optional, Union, cast
+from tree import Result, Node
+from query import Query, QueryStateDict
+from queries import AnswerTuple, LatLonTuple, MONTH_ABBREV, gen_answer
+
+import datetime
+import random
+import re
+import requests
+from bs4 import BeautifulSoup  # type: ignore
+from cachetools import cached, TTLCache
+from settings import changedlocale
+from geo import (
+    distance,
+    in_iceland,
+    capitalize_placename,
+    ICE_PLACENAME_BLACKLIST,
+)
+from iceaddr import placename_lookup  # type:ignore
+
+
+# Indicate that this module wants to handle parse trees for queries,
+# as opposed to simple literal text strings
+HANDLE_TREE = True
+
+# The grammar nonterminals this module wants to handle
+QUERY_NONTERMINALS = {"QSunQuery"}
+
+_SUN_QTYPE = "SunPosition"
+
+TOPIC_LEMMAS = [
+    "birting",
+    "dagsetur",
+    "dögun",
+    "hádegi",
+    "miðnætti",
+    "myrkur",
+    "sólarhæð",
+    "sólarlag",
+    "sólarupprás",
+    "sólris",
+    "sólsetur",
+    "sól",
+]
+
+
+def help_text(lemma: str) -> str:
+    """Help text to return when query.py is unable to parse a query but
+    one of the above lemmas is found in it"""
+    return "Ég get svarað ef þú spyrð til dæmis: {0}?".format(
+        random.choice(
+            (
+                "Hvenær reis sólin í morgun",
+                "Hvenær er sólsetur í kvöld",
+                "Hvenær rís sólin á morgun",
+            )
+        )
+    )
+
+
+# The context-free grammar for the queries recognized by this plug-in module
+GRAMMAR = """
+
+Query →
+    QSunQuery '?'?
+
+QSunQuery →
+    # Hvenær er birting    í Reykjavík á morgun
+    # Klukkan hvað var myrkur á Norðfirði í gær
+    QSunWhen QSunPositions QSunPlaceAndTime?
+    | QSunSunheight QSunPlaceAndTime?
+
+QSunPlaceAndTime →
+    QSunDate QSunLocation?
+    > QSunLocation QSunDate
+    > QSunLocation
+
+QSunWhen →
+    "hvenær" | "klukkan" "hvað"
+
+QSunIsWillWas →
+    'vera' | 'verða'
+
+
+QSunSunheight →
+    # TODO: "hver er hæð sólar í dag" isn't interpreted correctly
```

Comment:
This can be addressed through query module prioritization - there is a mechanism (a bit convoluted though) for this.